diff -up boost_1_48_0/boost/math/tools/config.hpp\~ boost_1_48_0/boost/math/tools/config.hpp
--- boost_1_48_0/boost/math/tools/config.hpp~	2011-10-16 20:57:30.000000000 +0200
+++ boost_1_48_0/boost/math/tools/config.hpp	2012-01-25 01:50:44.446832172 +0100
@@ -24,7 +24,7 @@
 #include <boost/math/special_functions/detail/round_fwd.hpp>
 
 #if (defined(__CYGWIN__) || defined(__FreeBSD__) || defined(__NetBSD__) \
-   || (defined(__hppa) && !defined(__OpenBSD__)) || defined(__NO_LONG_DOUBLE_MATH)) && !defined(BOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS)
+   || (defined(__hppa) && !defined(__OpenBSD__))) && !defined(BOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS)
 #  define BOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS
 #endif
 #if BOOST_WORKAROUND(__BORLANDC__, BOOST_TESTED_AT(0x582))

Diff finished.  Wed Jan 25 01:50:49 2012
