diff -uNr deltarpm-git-20090913.orig/Makefile deltarpm-git-20090913/Makefile
--- deltarpm-git-20090913.orig/Makefile	2013-11-26 12:03:52.648684083 +0700
+++ deltarpm-git-20090913/Makefile	2013-11-26 12:04:13.169832403 +0700
@@ -24,7 +24,7 @@
 combinedeltarpm: combinedeltarpm.o md5.o util.o rpmhead.o cfile.o readdeltarpm.o writedeltarpm.o $(zlibbundled)
 
 rpmdumpheader: rpmdumpheader.o
-	$(CC) $(LDFLAGS) $^ -lrpm -o $@
+	$(CC) $(LDFLAGS) $^ -lrpm -lrpmio -o $@
 
 makedeltaiso: makedeltaiso.o delta.o rpmoffs.o rpmhead.o util.o md5.o cfile.o $(zlibbundled)
 
