diff -uNr gcc-4.6.3-20120301.orig/config-ml.in gcc-4.6.3-20120301/config-ml.in
--- gcc-4.6.3-20120301.orig/config-ml.in	2010-11-20 03:25:49.658795000 +0700
+++ gcc-4.6.3-20120301/config-ml.in	2013-11-21 20:20:34.025844196 +0700
@@ -516,6 +516,8 @@
 				ADAFLAGS="$(ADAFLAGS) $${flags}" \
 				prefix="$(prefix)" \
 				exec_prefix="$(exec_prefix)" \
+				mandir="$(mandir)" \
+				infodir="$(infodir)" \
 				GCJFLAGS="$(GCJFLAGS) $${flags}" \
 				GOCFLAGS="$(GOCFLAGS) $${flags}" \
 				CXXFLAGS="$(CXXFLAGS) $${flags}" \
diff -uNr gcc-4.6.3-20120301.orig/gcc/ada/sem_util.adb gcc-4.6.3-20120301/gcc/ada/sem_util.adb
--- gcc-4.6.3-20120301.orig/gcc/ada/sem_util.adb	2010-12-20 14:26:57.614932000 +0700
+++ gcc-4.6.3-20120301/gcc/ada/sem_util.adb	2013-11-21 20:18:40.110411685 +0700
@@ -2527,7 +2527,11 @@
          end if;
 
       elsif Is_Entity_Name (A2) then
+
+         --  Triggered by -Wall
+         pragma Warnings (Off);
          return Denotes_Same_Prefix (A2, A1);
+         pragma Warnings (On);
 
       elsif Nkind_In (A1, N_Selected_Component, N_Indexed_Component, N_Slice)
               and then
@@ -4967,7 +4971,7 @@
             Exp           : Node_Id;
             Assn          : Node_Id;
             Choice        : Node_Id;
-            Comp_Type     : Entity_Id;
+            Comp_Type     : Entity_Id := Empty;
             Is_Array_Aggr : Boolean;
 
          begin
diff -uNr gcc-4.6.3-20120301.orig/libada/Makefile.in gcc-4.6.3-20120301/libada/Makefile.in
--- gcc-4.6.3-20120301.orig/libada/Makefile.in	2012-03-06 16:40:54.402394000 +0700
+++ gcc-4.6.3-20120301/libada/Makefile.in	2013-11-21 20:11:42.027155733 +0700
@@ -72,18 +72,40 @@
 ADA_RTS_DIR=$(GCC_DIR)/ada/rts$(subst /,_,$(MULTISUBDIR))
 ADA_RTS_SUBDIR=./rts$(subst /,_,$(MULTISUBDIR))
 
+DEFAULTMULTIFLAGS :=
+ifeq ($(MULTISUBDIR),)
+targ:=$(subst -, ,$(target))
+arch:=$(word 1,$(targ))
+ifeq ($(words $(targ)),2)
+osys:=$(word 2,$(targ))
+else
+osys:=$(word 3,$(targ))
+endif
+ifeq ($(strip $(filter-out i%86 x86_64 powerpc% ppc% s390% sparc% linux%, $(arch) $(osys))),)
+ifeq ($(shell $(CC) $(CFLAGS) -print-multi-os-directory),../lib64)
+DEFAULTMULTIFLAGS := -m64
+else
+ifeq ($(strip $(filter-out s390%, $(arch))),)
+DEFAULTMULTIFLAGS := -m31
+else
+DEFAULTMULTIFLAGS := -m32
+endif
+endif
+endif
+endif
+
 # exeext should not be used because it's the *host* exeext.  We're building
 # a *target* library, aren't we?!?  Likewise for CC.  Still, provide bogus
 # definitions just in case something slips through the safety net provided
 # by recursive make invocations in gcc/ada/Makefile.in
 LIBADA_FLAGS_TO_PASS = \
         "MAKEOVERRIDES=" \
-        "LDFLAGS=$(LDFLAGS)" \
+        "LDFLAGS=$(strip $(LDFLAGS) $(DEFAULTMULTIFLAGS))" \
         "LN_S=$(LN_S)" \
         "SHELL=$(SHELL)" \
-        "GNATLIBFLAGS=$(GNATLIBFLAGS) $(MULTIFLAGS)" \
-        "GNATLIBCFLAGS=$(GNATLIBCFLAGS) $(MULTIFLAGS)" \
-        "GNATLIBCFLAGS_FOR_C=$(GNATLIBCFLAGS_FOR_C) $(MULTIFLAGS)" \
+        "GNATLIBFLAGS=$(strip $(GNATLIBFLAGS) $(MULTIFLAGS) $(DEFAULTMULTIFLAGS))" \
+        "GNATLIBCFLAGS=$(strip $(GNATLIBCFLAGS) $(MULTIFLAGS) $(DEFAULTMULTIFLAGS))" \
+        "GNATLIBCFLAGS_FOR_C=$(strip $(GNATLIBCFLAGS_FOR_C) $(MULTIFLAGS) $(DEFAULTMULTIFLAGS))" \
         "TARGET_LIBGCC2_CFLAGS=$(TARGET_LIBGCC2_CFLAGS)" \
         "THREAD_KIND=$(THREAD_KIND)" \
         "TRACE=$(TRACE)" \
@@ -94,7 +116,7 @@
         "exeext=.exeext.should.not.be.used " \
 	'CC=the.host.compiler.should.not.be.needed' \
 	"GCC_FOR_TARGET=$(CC)" \
-        "CFLAGS=$(CFLAGS) $(WARN_CFLAGS)"
+        "CFLAGS=$(strip $(CFLAGS) $(DEFAULTMULTIFLAGS) $(WARN_CFLAGS))"
 
 # Rules to build gnatlib.
 .PHONY: gnatlib gnatlib-plain gnatlib-sjlj gnatlib-zcx gnatlib-shared osconstool
diff -uNr gcc-4.6.3-20120301.orig/libjava/Makefile.am gcc-4.6.3-20120301/libjava/Makefile.am
--- gcc-4.6.3-20120301.orig/libjava/Makefile.am	2010-08-27 11:27:40.775868000 +0700
+++ gcc-4.6.3-20120301/libjava/Makefile.am	2013-11-21 20:23:38.196160534 +0700
@@ -755,7 +755,8 @@
 ## later.
 	@echo Installing dummy lib libgcj_bc.so.1.0.0; \
 	rm $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so; \
-	mv $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so.1.0.0 $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so; \
+	$(INSTALL) $(INSTALL_STRIP_FLAG) $(here)/.libs/libgcj_bc.so $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so; \
+	rm $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so.1.0.0; \
 	$(libgcj_bc_dummy_LINK) -xc /dev/null -Wl,-soname,libgcj_bc.so.1 \
 	-o $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so.1.0.0 -lgcj || exit; \
 	rm $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so.1; \
diff -uNr gcc-4.6.3-20120301.orig/libjava/Makefile.in gcc-4.6.3-20120301/libjava/Makefile.in
--- gcc-4.6.3-20120301.orig/libjava/Makefile.in	2011-02-05 15:32:56.477225000 +0700
+++ gcc-4.6.3-20120301/libjava/Makefile.in	2013-11-21 20:29:09.830325611 +0700
@@ -12401,7 +12401,8 @@
 	install-libexecsubPROGRAMS
 @USE_LIBGCJ_BC_TRUE@	@echo Installing dummy lib libgcj_bc.so.1.0.0; \
 @USE_LIBGCJ_BC_TRUE@	rm $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so; \
-@USE_LIBGCJ_BC_TRUE@	mv $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so.1.0.0 $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so; \
+@USE_LIBGCJ_BC_TRUE@	$(INSTALL) $(INSTALL_STRIP_FLAG) $(here)/.libs/libgcj_bc.so $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so; \
+@USE_LIBGCJ_BC_TRUE@	rm $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so.1.0.0; \
 @USE_LIBGCJ_BC_TRUE@	$(libgcj_bc_dummy_LINK) -xc /dev/null -Wl,-soname,libgcj_bc.so.1 \
 @USE_LIBGCJ_BC_TRUE@	-o $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so.1.0.0 -lgcj || exit; \
 @USE_LIBGCJ_BC_TRUE@	rm $(DESTDIR)$(toolexeclibdir)/libgcj_bc.so.1; \
