diff -uNr gcc-4.6.3-20120301.orig/libjava/prims.cc gcc-4.6.3-20120301/libjava/prims.cc
--- gcc-4.6.3-20120301.orig/libjava/prims.cc	2012-03-06 16:40:54.402394000 +0700
+++ gcc-4.6.3-20120301/libjava/prims.cc	2013-11-21 21:08:55.082170623 +0700
@@ -38,6 +38,14 @@
 #endif
 
 #ifndef DISABLE_GETENV_PROPERTIES
+#ifdef __GLIBC__
+/* glibc 2.15+ provides even for C++ inline optimized ::isspace etc.
+   Unfortunately those inlines are throw (), and call a function pointer
+   (which is throw () too, but with -fnon-call-exceptions this results
+   in a __cxa_call_unexpected call.  This macro disables the optimized
+   version.  */
+#define __NO_CTYPE 1
+#endif
 #include <java-props.h>
 #define PROCESS_GCJ_PROPERTIES process_gcj_properties()
 #else
